tests/quantum_locks_tests.cpp

```diff
@@ -351,6 +354,150 @@ TEST(ReadWriteSpinLock, Guards)
+    EXPECT_TRUE(lock.isReadLocked());
+    EXPECT_FALSE(lock.tryUpgradeToWrite()); //2 readers is impossible
+    EXPECT_FALSE(lock.isWriteLocked());
+    lock.unlockRead(); //only a single reader left
+    EXPECT_EQ(1, lock.numReaders());
+    
+    //start a bunch of parallel readers which will then upgrade to writers
+    std::vector<std::thread> threads;
+    auto timeToWake = std::chrono::system_clock::now() + std::chrono::milliseconds(10);
+    std::atomic_int count{0};
+    for (int i = 0; i < 10; ++i) {
+        threads.emplace_back([&lock, &timeToWake, &count]() {
+            lock.lockRead();
+            ++count;
+            std::this_thread::sleep_until(timeToWake);
+            lock.upgradeToWrite();
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+            EXPECT_GE(lock.numPendingWriters(), 0);
+            EXPECT_TRUE(lock.isWriteLocked());
+            lock.unlockWrite();
+        });
+    }
+    while (count < 10) {
+        //wait for threads to lock read
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    }
+    lock.upgradeToWrite(); //this will block waiting for the other threads to finish
+    EXPECT_TRUE(lock.isWriteLocked());
+    lock.unlockWrite();
+    for (auto&& t : threads) {
+        t.join();
+    }
+    EXPECT_EQ(0, lock.numReaders());
+    EXPECT_EQ(0, lock.numPendingWriters());
+    EXPECT_FALSE(lock.isLocked());
+}
+
+TEST(ReadWriteSpinLock, UpgradingBlockedMultipleReaders)
+{
+    std::vector<int> values;
+    ReadWriteSpinLock lock;
+    lock.lockRead();
+    lock.lockRead();
+    
+    //start a bunch of parallel readers which will then upgrade to writers
+    std::vector<std::thread> threads;
+    auto timeToWake = std::chrono::system_clock::now() + std::chrono::milliseconds(10);
+    std::atomic_int count{0};
+    for (int i = 0; i < 10; ++i) {
+        threads.emplace_back([&lock, &timeToWake, &values, &count, i]() {
+            if (i == 9) {
+                while (count < 9) {
+                    //make sure all other readers are blocked
+                    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+                }
+                lock.unlockRead(); //unblock the pending writer
+            }
+            else {
+                std::this_thread::sleep_until(timeToWake);
+                ++count;
+                lock.lockRead(); //this will block until the writer upgrades
+                lock.upgradeToWrite();
+                values.push_back(i);
+                lock.unlockWrite();
+            }
+        });
+    }
+    //this will block waiting for the other threads to finish
+    lock.upgradeToWrite();
+    EXPECT_TRUE(lock.isWriteLocked());
+    values.push_back(-1);
+    lock.unlockWrite();
+    for (auto&& t : threads) {
+        t.join();
+    }
+    EXPECT_EQ(0, lock.numReaders());
+    EXPECT_EQ(0, lock.numPendingWriters());
+    EXPECT_FALSE(lock.isLocked());
+    EXPECT_EQ(-1, values.front());
+}
+
```

Comment:
Please also add tests of the guards, especially the ReadGuard upgrading to a write lock and correctly unlocking.